card/db/migrate/20190116170824_add_left_key_to_card_virtuals.rb

```diff
@@ -0,0 +1,11 @@
+# -*- encoding : utf-8 -*-
+
+class AddLeftKeyToCardVirtuals < ActiveRecord::Migration[4.2]
+  def up
+    add_column :card_virtuals, :left_key, :string
+  end
+
+  def down
+    remove_column :card_virtuals, :left_key
+  end
+end
```

Comment:
It's a standard rails migration and not a card migration. There are rollback and remigrate tasks that expect that method.

Reply:
I meant I wasn't clear on how we're using the `card_virtuals` table in general.  (but I admit that wasn't clear!)

Reply:
We use it to cache the content of virtual cards. We used to use left_id and right_id as identifier but that doesn't work for virtuals cards whose left part doesn't exist. *machine_cache cards are an example. Because left_id is sometimes nil for those cards and all have the same right_id some machine input cards shared the same cached content.